src/docfx/build/xref/XrefMap.cs

```diff
@@ -354,21 +358,105 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
             }
 
             var errors = new List<Error>();
-            var (schemaErrors, content) = JsonUtility.ToObjectWithSchemaValidation(
+            var (schemaErrors, _) = JsonUtility.ToObjectWithSchemaValidation(
                 obj,
                 schema.Type,
                 transform: AttributeTransformer.TransformXref(context, file, null, extensionData));
 
             errors.AddRange(schemaErrors);
-            var xref = new InternalXrefSpec
+
+            var extensionDataByUid = new Dictionary<string, (bool isRoot, Dictionary<string, Lazy<JValue>> properties)>();
+
+            if (extensionData.Count == 0)
             {
-                Uid = uid,
-                Href = file.CanonicalUrlWithoutLocale,
-                ReferencedFile = file,
-            };
+                foreach (var (uid, value) in uids)
+                {
+                    extensionDataByUid.Add(uid, (string.IsNullOrEmpty(value), new Dictionary<string, Lazy<JValue>>()));
+                }
+            }
+            else
+            {
+                foreach (var (jsonPath, xrefProperty) in extensionData)
+                {
+                    var (uid, resolvedJsonPath) = MatchExtensionDataToUid(jsonPath);
+                    if (extensionDataByUid.ContainsKey(uid))
+                    {
+                        var (_, properties) = extensionDataByUid[uid];
+                        properties.Add(resolvedJsonPath, xrefProperty);
+                    }
+                    else
+                    {
+                        extensionDataByUid.Add(uid, (string.IsNullOrEmpty(uids[uid]), new Dictionary<string, Lazy<JValue>> { { resolvedJsonPath, xrefProperty } }));
+                    }
+                }
+            }
 
-            xref.ExtensionData.AddRange(extensionData);
-            return (errors, xref);
+            var specs = extensionDataByUid.Select(item =>
+            {
+                var (isRoot, properties) = item.Value;
+                var xref = new InternalXrefSpec
+                {
+                    Uid = item.Key,
+                    Href = isRoot ? file.CanonicalUrlWithoutLocale : $"{file.CanonicalUrlWithoutLocale}#{GetBookmarkFromUid(item.Key)}",
+                    ReferencedFile = file,
+                };
+                xref.ExtensionData.AddRange(properties);
+                return xref;
+            }).ToList();
+
+            return (errors, specs);
+
+            string GetBookmarkFromUid(string uid)
+                => Regex.Replace(uid, @"\W", "_");
+
+            (string uid, string jsonPath) MatchExtensionDataToUid(string jsonPath)
+            {
+                var reverse = uids.ToDictionary(x => x.Value, x => x.Key);
```

Comment:
build the `reverse` lookup upfront to save duplicated work here. Name them something like `uidToJsonPath` and `jsonPathToUid` so we know what each `Dict<string, string>` represents

Reply:
sounds good